AA-614: Add initial effort estimation block transformer

mikix opened the pull request:
https://openedx.atlassian.net/browse/AA-614

mikix commented on openedx/features/course_experience/templates/course_experience/course-outline-fragment.html:
```diff
@@ -50,6 +50,7 @@
                             id="${ section['id'] }">
                         <span class="fa fa-chevron-right ${ 'fa-rotate-90' if section_is_auto_opened else '' }" aria-hidden="true"></span>
                         <h3 class="section-title">${ section['display_name'] }</h3>
+                            <span>- <b>Estimate: ${section.get('effort_time', None)}, ${section.get('effort_activities', None)}</b></span>
```
This is just for testing - will be removed.

mikix commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,209 @@
+        # this course.
+        duration = max(0, cache[cls.CACHE_VIDEO_DURATIONS].get(xblock.edx_video_id, 0))
+        if not duration:
+            print("MIKE did not find video duration info", xblock.location, xblock.edx_video_id)
+            # FIXME raise cls.MissingEstimationData()
+            duration = 360
```
Obviously, this is here just for testing atm.

mikix commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -102,15 +103,16 @@ def get_blocks(
         )
     ]
 
-    if include_completion:
-        transformers += [BlockCompletionTransformer()]
```
This change is unrelated to my actual purpose.  I just noticed that we were manually reproducing the work of a kwarg to `get_course_blocks` - so figured I'd not duplicate the code.

MatthewPiatetsky commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+
+class EffortEstimationTransformer(BlockStructureTransformer):
+    """
+    A transformer that adds a partition condition for all graded content
```
is this comment accurate? 
Also could we include in the docstring a description of what the feature is and perhaps a link to a product doc on it and there is one that is publicly available?

mikix replied:
Ah yeah just a copy / paste error. Will update.

MatthewPiatetsky commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+    READ_VERSION = 1
+
+    # Public xblock field names
+    EFFORT_ACTIVITIES = 'effort_activities'
```
these are set to effort_activities and effort_time here, but elsewhere in your PR you're using those hardcoded values.  
Should those other spots be importing from here?

mikix replied:
I used these class properties in a couple more places where we already had a reference to this class. But in general, I'd like to avoid importing this random transformer just to grab a string from it. The string can't really change anyway.

MatthewPiatetsky commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+        """
+        Grabs raw estimates for leaf content.
+
+        Pooled estimates higher on verticals and such is done by transform() below, because what each user sees could
```
could you reword this comment, i'm having a hard time parsing what it means

MatthewPiatetsky commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+
+        # Check if we have a duration. If not, raise an exception that will stop this transformer from affecting
+        # this course.
+        duration = max(0, cache[cls.CACHE_VIDEO_DURATIONS].get(xblock.edx_video_id, 0))
```
what is this returning if it isn't in the cache? wonder if there is a clearer way to write this

mikix replied:
In the `get(xblock.edx_video_id, 0)` bit, the `0` is the default value for a dictionary lookup. So `0` gets returned if it's not in the cache.

Do you know what about the line is not clear? Is the `max(0, ...)` bit making it harder to read?

MatthewPiatetsky replied:
more of a nit so you can feel free to ignore but you could do something like
`cached_duration = cache[cls.CACHE_VIDEO_DURATIONS].get(xblock.edx_video_id)
duration = cached_duration if cached_duration else 0`

MatthewPiatetsky commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+        # Get bucket for this experiment. 0 is no estimate. 1 is only on sections. 2 is only on subsections.
+        # If we drop the experiment but keep the feature - we should leave estimates on all blocks, and just have the
+        # UI show the estimate at the level they care about.
+        bucket = EFFORT_ESTIMATION_LOCATION_FLAG.get_bucket(course_key=usage_info.course_key)
```
are we bucketing on the course and not on the user?  
not sure if we have a test plan written up, but I would have expected the bucketing to me on a user level with randomly assigned buckets

mikix replied:
We always bucket by user. There's no way not to. Waffle flag code (and our experiment flag code built on top of it) goes and gets the current request to get the user under consideration (this is one reason why we reset `request.user` when masquerading).

But the `course_key=` bit means we *also* bucket per course (that is, per-enrollment, rather than per-user). So rather than being bucketed into one experience across edx.org, you may see different experiences in each of your courses. I asked Tim which way he wanted it.

MatthewPiatetsky replied:
ah ok, when i saw the course_key argument i misunderstood, thanks for clarifying!

MatthewPiatetsky commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+
+        # Second pass to clear out collected estimate on levels we don't want to share. Just an easy way to test
+        # estimates at different levels, per the experiment. If we make this real code and not an experiment, just
+        # drop this whole section.
```
should there be a TODO of some sort here?  
or perhaps we make a ticket of things to address if we make this code real and put the ticket number here?

mikix replied:
I made https://openedx.atlassian.net/browse/AA-659 to hold that stuff and added a couple references to it around where I thought it might help.

MatthewPiatetsky commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+        be different.
+        """
+        block_structure.request_xblock_fields('category')
+        block_structure.request_xblock_fields('global_speed', 'only_on_web')  # video fields
```
what is global speed?

mikix replied:
`global_speed` is the user's preference for the default video speed. `speed` is the last-used speed for a given video. We only look at `global_speed` in this transformer.

MatthewPiatetsky reviewed (commented):
assuming no tests because it is an experiment?
i think we'd want to have some more extensive documentation on what this feature is, but again perhaps that's something that would be postponed for if/when this was implemented for real

so i guess if we're taking the approach of "this is an experiment so we'll just put together whatever works and make it nice and clean later" then all of my comments can probably be safely ignored

Dillon-Dumesnil commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+            for block_key in block_structure.topological_traversal():
+                xblock = block_structure.get_xblock(block_key)
+
+                collections = {
+                    'html': cls._collect_html_effort,
+                    'video': cls._collect_video_effort,
+                }
```
Define this outside of the for loop. Doesn't look like we need to restart it each loop.

Dillon-Dumesnil commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,225 @@
+            category = block_structure.get_xblock_field(block_key, 'category')
+
+            # These estimation methods should return a tuple of (a number in seconds, an activity count)
+            estimations = {
+                'chapter': self._estimate_children_effort,
+                'course': self._estimate_children_effort,
+                'html': self._estimate_html_effort,
+                'sequential': self._estimate_children_effort,
+                'vertical': self._estimate_vertical_effort,
+                'video': self._estimate_video_effort,
+            }
```
Similarly here, can this be moved outside of the for loop?

Dillon-Dumesnil reviewed (commented):
Followed what I could, but yeah, looked pretty fine to me. Reiterate Matthew's comments and added some small ones of my own.

mikix commented:
> assuming no tests because it is an experiment?

My plan is to add some tests, but not as extensive as we should do. It's an experiment, but it's also still a block transformer that will always run (even if it's often a no-op for learner). So I don't feel comfortable doing *no* tests.

mikix commented:
I made https://openedx.atlassian.net/browse/AA-659 to track the follow-on cleanup work.

mikix commented:
OK, I *think* I addressed all comments (barring the lack of tests, still gonna work on that in a sec).

Functional changes since you last reviewed:
- Made the experiment flag non-course-aware (so a user is in the same bucket across all their courses). This was at Tim's request.
- Delay bucketing until we see if the course even has any raw estimate data to use. This will avoid bucketing folks until we get a chance to re-publish courses which fills in the data.

MatthewPiatetsky commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,239 @@
+    def _estimate_children_effort(self, _usage_info, block_structure, block_key):
+        """Collects time and activity counts for children."""
+        time = self._gather_child_values(block_structure, block_key, self.EFFORT_TIME)
+        time = time or None  # avoid claiming anything takes 0 seconds by coercing to None (no estimate) instead
```
nit Why not just return None from the children?

mikix replied:
I don't think it ended up being a practical difference - some versions of my PR might have treated them differently, can't remember. But mostly because I felt it was more appropriate to say "unknown" than "zero" if we had no time info? Very minor thing.

MatthewPiatetsky reviewed (approved):
:shipit:

mikix commented on openedx/features/effort_estimation/block_transformers.py:
```diff
@@ -0,0 +1,239 @@
+        # Some videos will suggest specific start & end times, rather than the whole video. Note that this is only
+        # supported in some clients (other clients - like the mobile app - will play the whole video anyway). So we
+        # record this duration separately, to use instead of the whole video duration if the client supports it.
+        clip_duration = (xblock.end_time - xblock.start_time).total_seconds()
```
Yeah

mikix replied:
Oh sorry, they are `datetime.timedelta` objects, not datetimes. I lied to you. Point was that they are time objects that can be subtracted etc.
